pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		if configID == dynamicRatesByService || configID == userDefinedRatesByService {
+			var payload apmsampling.APMSampling
+			_, err := payload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal APMSampling remote config envPayload: %w", err)
+			}
+			remoteRateUpdates = append(remoteRateUpdates, sampler.RemoteRateUpdate{Version: conf.Metadata.Version, Config: payload})
+		}
+		if configID == samplerconfig {
+			samplerconfigPayload = &apmsampling.SamplerConfig{}
+			_, err := samplerconfigPayload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal SamplerConfig remote config envPayload: %w", err)
+			}
+		}
+	}
+	return remoteRateUpdates, samplerconfigPayload, nil
+}
+
+func (h *RemoteConfigHandler) configPathToID(path string) (apmsamplingRemoteConfigID, error) {
+	pathMatch := h.remoteConfigPathRegex.FindStringSubmatch(path)
+	if pathMatch == nil {
+		return "", fmt.Errorf("failed to match remote config path %s", path)
+	}
+
+	switch pathMatch[1] {
+	case string(dynamicRatesByService):
+		return dynamicRatesByService, nil
+	case string(userDefinedRatesByService):
+		return userDefinedRatesByService, nil
+	case string(samplerconfig):
+		return samplerconfig, nil
+	default:
+		return "", fmt.Errorf("unexpected remote config ID: %s", pathMatch[1])
+	}
+}
+
+func (h *RemoteConfigHandler) updateSamplers(config apmsampling.SamplerConfig) {
+	var confForEnv *apmsampling.SamplerEnvConfig
+	for _, envAndConfig := range config.ByEnv {
+		if envAndConfig.Env == h.agentConfig.DefaultEnv {
+			confForEnv = &envAndConfig.Config
+		}
+	}
+
+	var prioritySamplerTargetTPS float64
+	if confForEnv != nil && confForEnv.PrioritySamplerTargetTPS != nil {
```

Comment:
I feel like there's a good amount of duplication here but I can't come up with a good way to reduce it. I tried making a helper function but it ended up being more complicated than I wanted. Any thoughts or ideas here?

Reply:
One idea I had was you could pull the final `else` to the top line by initializing each var with their default value like:
`prioritySamplerTargetTPS := h.agentConfig.TargetTPS` (which ends up saving 9 lines total)

Reply:
We have 3 branches here, and even a good refactoring couldn't go lower than 2 branches since there are the "float" and the "bool" cases.

So my idea was it was probably not worth introducing the overhead of an abstraction for such a small gain.

> One idea I had was you could pull the final else to the top line by initializing each var with their default value like:
> prioritySamplerTargetTPS := h.agentConfig.TargetTPS (which ends up saving 9 lines total)

It does work. What I like about the current way it's written is that we go from highest precedence to lowest precedence in order, which is subjectively nice.

Of course everything would work here and I'm ok changing it :)